src/Type.h

```diff
@@ -142,20 +142,46 @@ template<> struct halide_c_type_to_name<float> { static const bool known_type =
 template<> struct halide_c_type_to_name<double> { static const bool known_type = true; static halide_cplusplus_type_name name() { return { halide_cplusplus_type_name::Simple,  "double"}; } };
 template<> struct halide_c_type_to_name<struct buffer_t> { static const bool known_type = true; static halide_cplusplus_type_name name() { return { halide_cplusplus_type_name::Struct,  "buffer_t"}; } };
 
+// Default case (should be only Unknown types, since we specialize for Known types below).
+// We require that all unknown types be pointers, and translate them all to void*
+// (preserving const-ness and volatile-ness).
 template<typename T, bool KnownType>
 struct halide_internal_handle_traits {
-    static const halide_handle_cplusplus_type *type_info(bool is_ptr, halide_handle_cplusplus_type::ReferenceType ref_type) { return nullptr; }
+    static const halide_handle_cplusplus_type *type_info(bool is_ptr, 
+            halide_handle_cplusplus_type::ReferenceType ref_type) { 
+        static_assert(!KnownType, "Only unknown types handled here");
+        internal_assert(is_ptr) << "Unknown types must be pointers";
+        internal_assert(ref_type == halide_handle_cplusplus_type::NotReference) << "Unknown types must not be references";
+        static const halide_handle_cplusplus_type the_info{
```

Comment:
I think this may be problematic as it is a static in a template in a header and we've run into issues with this before. (In particular it won't guarantee it is always the same pointer, but there were clang compilation issues I think too.) I would like type equivalence to be determinable by comparing pointers. (And I believe this is somewhat a requirement as it is probably possible for two distinct types to have different names. It certainly is in separate compilation contexts, though it gets messy fast.)

One possible solution is to call out to a non-templated routine defined in Type.cpp which returns a pointer to one of two file scope static halide_handle_cplusplus_type objects based on a bool param. We're now into enough cycles to construct a Type that I might think twice, but Handle types are rare so I don't think it matters.